assistant/assistant.go

```diff
@@ -0,0 +1,186 @@
+package assistant
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/influxdata/telegraf/agent"
+	"github.com/influxdata/telegraf/internal"
+)
+
+/*
+Assistant is a client to facilitate communications between Agent and Cloud.
+*/
+type Assistant struct {
+	config     *AssistantConfig // stores plugins
+	connection *websocket.Conn  // Active websocket connection
+	ctx        context.Context  // go's context
+	done       chan bool        // Channel used to stop server listener
+	agent      *agent.Agent     // Pointer to agent to issue commands
+}
+
+/*
+AssistantConfig allows us to configure where to connect to and other params
+for the agent.
+*/
+type AssistantConfig struct {
+	Host          string
+	Path          string
+	RetryInterval int
+}
+
+func (astConfig *AssistantConfig) fillDefaults() {
+	if astConfig.Host == "" {
+		astConfig.Host = "localhost:8080"
+	}
+	if astConfig.Path == "" {
+		astConfig.Path = "/echo"
+	}
+	if astConfig.RetryInterval == 0 {
+		astConfig.RetryInterval = 15
+	}
+}
+
+// NewAssistant returns an Assistant for the given Config.
+func NewAssistant(ctx context.Context, config *AssistantConfig, agent *agent.Agent) (*Assistant, error) {
+	config.fillDefaults()
+	var addr = flag.String("addr", config.Host, "http service address")
+	u := url.URL{Scheme: "ws", Host: *addr, Path: config.Path}
+
+	header := http.Header{}
+
+	if v, exists := os.LookupEnv("INFLUX_TOKEN"); exists {
+		header.Add("Authorization", "Token "+v)
+	} else {
+		return nil, fmt.Errorf("Influx authorization token not found, please set in env")
+	}
+
+	// creates a new websocket connection
+	log.Printf("D! [assistant] Attempting connection to [%s]", config.Host)
+	ws, _, err := websocket.DefaultDialer.Dial(u.String(), header)
+	for err != nil { // on error, retry connection again
+		log.Printf("E! [assistant] Failed to connect to [%s], retrying in %ds, "+
+			"error was '%s'", config.Host, config.RetryInterval, err)
+
+		err := internal.SleepContext(ctx, time.Duration(config.RetryInterval)*time.Second)
+		if err != nil {
+			return nil, err
+		}
+
+		ws, _, err = websocket.DefaultDialer.Dial(u.String(), header)
+	}
+	log.Printf("D! [assistant] Successfully connected to %s", config.Host)
+
+	a := &Assistant{
+		config:     config,
+		connection: ws,
+		ctx:        ctx,
+		done:       make(chan bool),
+		agent:      agent,
+	}
+
+	go a.listenToServer()
+
+	return a, nil
+}
+
+// Stop is used to clean up active connection and all channels
+func (assistant *Assistant) Stop() {
+	assistant.connection.Close()
+	assistant.done <- true
```

Comment:
nit: Not sure if this is true convention (you can look it up), but from my experience normally when you're just using a channel to send signals and not actually doing anything with any values, you normally use a generic struct{} chan. I don't think this is necessarily a big deal at all, but like the first thought when I saw ```assistant.done <- true``` was to go look for what it was doing with the true value, and if there was anywhere in the code it was sending in false.

Unless there's a plan to send in multiple values and handle them differently, it might be better to stick with a more generic channel:
```
done: make(chan struct{})

done <- struct{}{}
```